res/controllers/Ion-Discover-DJ-Pro-scripts.js

```diff
@@ -0,0 +1,262 @@
+IonDiscoverDjPro.setLED = function(value, status) {
+    if (status) {
+        status = 0x64;
+    } else {
+        status = 0x00;
+    }
+    midi.sendShortMsg(0x90, value, status);
+};
+
+IonDiscoverDjPro.samplesPerBeat = function(group) {
+    // FIXME: Get correct samplerate and channels for current deck
+    var sampleRate = 44100;
+    var channels = 2;
+    var bpm = engine.getValue(group, "file_bpm");
+    return channels * sampleRate * 60 / bpm;
+};
+
+IonDiscoverDjPro.selectKnob = function(channel, control, value, status, group) {
+    if (value > 63) {
+        value = value - 128;
+    }
+    if (IonDiscoverDjPro.directoryMode) {
+        var i;
+        if (value > 0) {
+            for (i = 0; i < value; i++) {
+                engine.setValue(group, "SelectNextPlaylist", 1);
+            }
+        } else {
+            for (i = 0; i < -value; i++) {
+                engine.setValue(group, "SelectPrevPlaylist", 1);
+            }
+        }
+    } else {
+        engine.setValue(group, "SelectTrackKnob", value);
+    }
+};
+
+IonDiscoverDjPro.loopIn = function(channel, control, value, status, group) {
+    var deck = IonDiscoverDjPro.groupToDeck(group);
+    if (value) {
+        if (IonDiscoverDjPro.manualLooping[deck-1]) {
+            // Cut loop to Half
+            var start = engine.getValue(group, "loop_start_position");
+            var end = engine.getValue(group, "loop_end_position");
+            if ((start !== -1) && (end !== -1)) {
+                var len = (end - start) / 2;
+                engine.setValue(group, "loop_end_position", start + len);
+            }
+        } else {
+            if (engine.getValue(group, "loop_enabled")) {
+                engine.setValue(group, "reloop_exit", 1);
+                IonDiscoverDjPro.setLED(IonDiscoverDjPro.leds[deck]["reLoop"], false);
+            }
+            engine.setValue(group, "loop_in", 1);
+            engine.setValue(group, "loop_end_position", -1);
+            IonDiscoverDjPro.setLED(IonDiscoverDjPro.leds[deck]["loopIn"], true);
+            IonDiscoverDjPro.setLED(IonDiscoverDjPro.leds[deck]["loopOut"], false);
+        }
+    }
+};
+
+IonDiscoverDjPro.loopOut = function(channel, control, value, status, group) {
+    var deck = IonDiscoverDjPro.groupToDeck(group);
+    if (value) {
+        var start = engine.getValue(group, "loop_start_position");
+        var end = engine.getValue(group, "loop_end_position");
+        if (IonDiscoverDjPro.manualLooping[deck-1]) {
+            // Set loop to current Bar (very approximative and would need to get fixed !!!)
+            var bar = IonDiscoverDjPro.samplesPerBeat(group);
+            engine.setValue(group, "loop_in", 1);
+            engine.setValue(group, "loop_end_position", start + bar);
+            IonDiscoverDjPro.setLED(IonDiscoverDjPro.leds[deck]["loopIn"], true);
+            IonDiscoverDjPro.setLED(IonDiscoverDjPro.leds[deck]["loopOut"], true);
+        } else {
+            if (start !== -1) {
```

Comment:
You can reduce nesting depth by using early-exit here:

```suggestion
            if (start === -1) {
                return;
            } 
```

Reply:
Done